### PR TITLE
[KYUUBI #4526][FOLLOWUP] Defer evaluation for GroupProvider

### PR DESCRIPTION
**Pull request by merrily01**
### _Why are the changes needed?_

As described in the discussion: https://github.com/apache/kyuubi/discussions/4512 

Since the pr https://github.com/apache/kyuubi/pull/3897 , we introduced the feature of `Supplying pluggable GroupProvider`.

But now when we create a new EngineRef, no matter which engine share level is used, getGroupNames will always be called.

This may cause users to encounter some confusion stack trace when they do not use the GROUP engine share level.

So we solve the problem through moving the `sessionManager.groupProvider.primaryGroup` to the inside of `EngineRef.appUser`, to ensure that the feature of `Supplying pluggable GroupProvider` is only valid for the GROUP engine share level.

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
@merrily01, thanks for your contribution, would you mind updating your GitHub profile to link the email <maruilei@58.com> to your GitHub account? Otherwise, your contributions will not be counted by GitHub.

**Comment by merrily01** [comment]
> @merrily01, thanks for your contribution, would you mind updating your GitHub profile to link the email [maruilei@58.com](mailto:maruilei@58.com) to your GitHub account? Otherwise, your contributions will not be counted by GitHub.

Sorry，already updated.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 1]
```diff
@@ -51,7 +53,7 @@ import org.apache.kyuubi.operation.log.OperationLog
 private[kyuubi] class EngineRef(
     conf: KyuubiConf,
     user: String,
-    primaryGroup: String,
+    groupProvider: GroupProvider,
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:56]: It seems also can be lazily initialized, not needed in the constructor?

Comment (merrily01) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:56]: Thanks for your reply.

It seems that the groupProvider needs to obtain from KyuubiSessionManager. Whether it is passing KyuubiSessionManager through the constructor or trying to initialize and create one, the chaotic changes introduced and the cost paid may be greater than the current transformation.

Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:56]: the cost is mostly test code refactoring, but I'm OK to keep the PR AS-IS.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala [hunk 2]
```diff
@@ -77,7 +77,7 @@ class KyuubiSessionImpl(
   lazy val engine: EngineRef = new EngineRef(
     sessionConf,
     user,
-    sessionManager.groupProvider.primaryGroup(user, optimizedConf.asJava),
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:80]: this is also a bugfix as using optimized conf is wrong here

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 3]
```diff
@@ -85,7 +87,7 @@ private[kyuubi] class EngineRef(
   // Launcher of the engine
   private[kyuubi] val appUser: String = shareLevel match {
     case SERVER => Utils.currentUser
-    case GROUP => primaryGroup
+    case GROUP => groupProvider.primaryGroup(user, Map.empty[String, String].asJava)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:90]: sessionConf.getAll

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefTests.scala [hunk 4]
```diff
@@ -68,7 +71,9 @@ trait EngineRefTests extends KyuubiFunSuite {
     Seq(None, Some("suffix")).foreach { domain =>
       conf.set(KyuubiConf.ENGINE_SHARE_LEVEL, CONNECTION.toString)
       domain.foreach(conf.set(KyuubiConf.ENGINE_SHARE_LEVEL_SUBDOMAIN.key, _))
-      val engine = new EngineRef(conf, user, "grp", id, null)
+      conf.set(KyuubiConf.GROUP_PROVIDER, "hadoop")
+
+      val engine = new EngineRef(conf, user, PluginLoader.loadGroupProvider(conf), id, null)
```
Comment (yaooqinn) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefTests.scala:76]: The group provider is unnecessary here, but OK to hold down the LOC

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala (hunk 1 above)
Comment (merrily01) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:56]: Ok, thank you

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala (hunk 2 above)
Comment (merrily01) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/session/KyuubiSessionImpl.scala:80]: Yeah, actually it seems that conf is not used in subsequent calls ? 
```
# HadoopGroupProvider.scala

  override def groups(user: String, sessionConf: JMap[String, String]): Array[String] =
    UserGroupInformation.createRemoteUser(user).getGroupNames match {
      case Array() =>
        warn(s"There is no group for $user, use the client user name as group directly")
        Array(user)
      case groups => groups
    }

```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala (hunk 3 above)
Comment (merrily01) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala]: Done

File: kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefTests.scala (hunk 4 above)
Comment (merrily01) [reply kyuubi-server/src/test/scala/org/apache/kyuubi/engine/EngineRefTests.scala:76]: Ok,  thank you

**Comment by merrily01** [comment]
Hi~ @yaooqinn @pan3793  sorry to bother you guys,

I found that some checks in the workflow failed, such as:

```
Error: Exception in thread "pool-182-thread-3" java.lang.Error: org.apache.kyuubi.KyuubiSQLException: Timeout to lock on path [/engine_test2_1.8.0-SNAPSHOT_USER_SPARK_SQL_lock/runner/default] after 4500 ms. There would be some problem that other session may create engine timeout.

```

And I don't know whether it is related to code change or test environment. 

So, mabey I want to know how to re-trigger the pipline?

In addition, by the way, in our project specification, when I modify or reply to the reviewer's comments, 

who should close the `Resolve conversation`? the contributor or reviewer?

**Comment by pan3793** [comment]
> ... whether it is related to code change or test environment.

@merrily01 it looks like a flaky test, @ulysses-you I remember you have fixed it but seems flaky again.

> who should close the Resolve conversation? the contributor or reviewer?

we don't have strict rules for that, normally, the committer will glance the conversations and evaluate if the PR can get merged.

**Pull request closed without merging**